flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adaptive/AdaptiveScheduler.java

```diff
@@ -265,7 +268,8 @@ public AdaptiveScheduler(
 
         this.executionGraphFactory = executionGraphFactory;
 
-        registerMetrics();
+        SchedulerBase.registerJobMetrics(
+                jobManagerJobMetricGroup, jobStatusStore, () -> (long) numRestarts);
```

Comment:
Does `numRestarts` need to be volatile? If I understand that correctly if we access the metric eg. using JMX, then it gets accessed by a different thread. Or is there some synchronization in the metrics system that I'm missing?

Reply:
> Does numRestarts need to be volatile?

Technically yes, but we generally don't do it. It would be too expensive on the hot code paths (aka, we can't be consistent about it anyway), and we haven't had issues so far 🤷 

> Or is there some synchronization in the metrics system that I'm missing?

There is none.